R/as-data-frame.R

```diff
@@ -31,51 +31,93 @@ NULL
 #' @rdname dataset-to-R
 #' @export
 as.data.frame.CrunchDataset <- function (x, row.names = NULL, optional = FALSE,
-                                        force=FALSE, categorical.mode = "factor",
-                                        include.hidden = FALSE,
-                                        row.order = NULL, ...) {
+    force=FALSE, categorical.mode = "factor",
+    include.hidden = FALSE,
+    row.order = NULL, ...) {
     out <- CrunchDataFrame(x, row.order = row.order,
-                           categorical.mode = categorical.mode,
-                           include.hidden = include.hidden)
+        categorical.mode = categorical.mode,
+        include.hidden = include.hidden)
     if (force) {
         out <- as.data.frame(out)
     }
     return(out)
 }
 
 #' @rdname dataset-to-R
+#' @importFrom utils read.csv
 #' @export
-as.data.frame.CrunchDataFrame <- function (x, row.names = NULL, optional = FALSE, ...) {
+as.data.frame.CrunchDataFrame <- function (x,
+    row.names = NULL,
+    optional = FALSE,
+    ...) {
     ds <- attr(x, "crunchDataset")
-    default.stringsAsFactors <- function () FALSE
-    limit <- min(c(10000, getOption("crunch.data.frame.limit")))
-    if (nrow(ds) * ncol(ds) > limit) {
-        ## TODO: switch to downloading CSV and reading that?
-        halt("Dataset too large to coerce to data.frame. ",
-            "Consider subsetting it first")
-    }
-    var_names <- names(x)
-    # TODO: something intelligent with modes
-    out <- lapply(var_names, function(var) {
-        values <- x[[var]]
-        
-        # Flatten all array variables, using only the subvariable name. We need 
-        # to create a list layer for non-dataframe values so that we can unlist 
-        # them later.
-        if (is.data.frame(values)) {
-            values <- as.list(values)
+    tmp <- tempfile()
+    csv_mode = ifelse(attr(x, "mode") == "factor", "name", "id")
+    write.csv(ds, tmp, categorical = csv_mode)
+    ds_out <- read.csv(tmp, stringsAsFactors = FALSE, colClasses = "character")
+    unlink(tmp)
+    return(csvToDataFrame(ds_out, x))
+}
+
+csvToDataFrame <- function(csv_df, cdf) {
```

Comment:
This is super minor, and I know it's internal so it's not a huge deal, but it would be good to use `crdf` instead of `cdf` here for consistency with our other abbreviations.